### DISCUSSION TITLE
Can we write a 2D array with dynamic shape?

### DISCUSSION THREAD

**Discussion by ziyuang** [BlueBrain/HighFive#767] [Questions] [answered]
I have a pointer to an array, and I know the shape. Can I write it as a 2D array? The following code will throw an exception of `Impossible to write buffer of dimensions [2, 3] into dataset with n = 1 dimensions.`:

```c++
float* raw = ...;
HighFive::DataSet dataset = file.createDataSet<float>("dataset", HighFive::DataSpace({2, 3}));
dataset.write(raw);
```

**Comment by ziyuang** [comment]
Closed: the solution is `dataset.write((float**)raw);`

**Reply by 1uc** [reply, marked as answer]
The downside of this is that when presented with a `float**` we can't know if it's contiguous or not, e.g.
```
float ** array = (float**) malloc(n * sizeof(float*));
for(size_t i = 0; i < n; ++i) {
    array[i] = malloc(m * sizeof(float));
}

// vs
float * data = (float*) malloc(n*m * sizeof(float));
float ** array = (float**) malloc(n*sizeof(float*));

for(size_t i = 0; i < n; ++i) {
  array[i] = data + i * n;
}
```

Currently, HighFive doesn't seem to respect the possibility of non-contiguous pointer-to-pointer. Hence, it might work. However, the version I posted should be preferred.

**Reply by ziyuang** [reply]
Right, makes sense. Thanks.

**Comment by 1uc** [comment]
There's `write_raw` which is probably the preferred solution in this case:
```
dataset.write_raw(raw);
```
See https://github.com/BlueBrain/HighFive/blob/master/src/examples/read_write_raw_ptr.cpp for a full example.

(Note that due to an asymmetry, there's no corresponding `read_raw`, it's just `read` with a pointer as argument.)
